docs/src/performance_tips.md

```diff
@@ -0,0 +1,105 @@
+# Performance tips and monitoring
+
+This document outlines some performance tips and performance monitoring strategies.
+
+For the most part, this document focuses on some of the common performance gotcha's that we've observed in the CliMA codebase.
+
+There is a very good and thorough overview of performance tips in [Julia's docs](https://docs.julialang.org/en/v1/manual/performance-tips/).
+
+## Avoiding global variables
+
+Julia allows for [function closures](https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-captured), which can be very handy, but can also result in performance cliffs, specifically if the captured variable is a non-constant global variable. So, for that reason, it's recommended to avoid closures when possible.
+
+## Dynamic memory allocations
+
+Some Julia functions incur allocations. For example, `push!` dynamically allocates memory. Sometimes, we can avoid using `push!` if the length of the container we're pushing to is known. If the length is unknown, then one can use alternative methods, for example, `map`. In addition, if `push!` is the only viable option, it's recommended to specify (if possible) the container type. For example, `Float64[]` and not `[]`. see [these docs](https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-abstract-container) for more details.
+
+## Tracking allocations
+
+Julia's performance docs above recommends to pay close attention to allocations. Allocations can be coarsely reported with the `@time` macro and more finely reported by using `julia --track-allocation=all`. From [CodeCov.jl's docs](https://github.com/JuliaCI/Coverage.jl#memory-allocation):
+
+Start julia with
+```sh
+julia --track-allocation=user
+```
+Then:
+- Run whatever commands you wish to test. This first run is to ensure that everything is compiled (because compilation allocates memory).
+- Call `Profile.clear_malloc_data()`
+- Run your commands again
+- Quit julia
+
+Finally, navigate to the directory holding your source code. Start julia (without command-line flags), and analyze the results using
+```julia
+using Coverage
+analyze_malloc(dirnames)  # could be "." for the current directory, or "src", etc.
+```
+This will return a vector of `MallocInfo` objects, specifying the number of bytes allocated, the file name, and the line number.
+These are sorted in increasing order of allocation size.
+
+## ReportMetrics.jl
+
+CliMA's [ReportMetrics.jl](https://github.com/CliMA/ReportMetrics.jl) applies the strategy in the above section and provides a re-useable interface for reporting the top-most important allocations. Here is an example of it in use:
+
+ - rep_workload.jl
+ - perf.jl
+
+```julia
+# File: rep_workload.jl
+import Profile
+
+import SpecialFunctions # for triggering invalidations
```

Comment:
Thanks Charlie for adding this! 
Just a minor question. What do you need this package for here?

Reply:
Ah, yeah, I can remove this, it was for something else

Reply:
That'd be nice. Thanks!